clang/lib/AST/ExprConstant.cpp

```diff
@@ -14574,9 +14574,17 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
   default:
     return false;
 
+  case Builtin::BI__builtin_frexpl:
+    // AIX library function `frexpl` has 'long double' type and not
+    // PPCDoubleDouble type. To make sure we generate the right value, don't
+    // constant evaluate it and instead defer to a libcall.
+    if (Info.Ctx.getTargetInfo().getTriple().isPPC() &&
+        (&Info.Ctx.getTargetInfo().getLongDoubleFormat() !=
+         &llvm::APFloat::PPCDoubleDouble()))
+      return false;
+    LLVM_FALLTHROUGH;
   case Builtin::BI__builtin_frexp:
-  case Builtin::BI__builtin_frexpf:
-  case Builtin::BI__builtin_frexpl: {
+  case Builtin::BI__builtin_frexpf: {
```

Comment:
The comment is incorrect. The _actual_ AIX library function `frexpl` has (some version of) `PPCDoubleDouble` type. Because of that, Clang (which only supports 64-bit `long double` on AIX) has to map calls to `__builtin_frexpl` to `frexp`.

For compile-time evaluation, the `APFloat`s associated with `long double` on AIX have the IEEE double format (and the evaluation works fine).

The solution to the failing test is to change it. The problem with the test is that the first input argument is a constant; changing it to refer to a global variable should work.
https://github.com/llvm/llvm-project/blob/566fbb450092bf8c9f966a6ab1b0381626e3e535/clang/test/CodeGen/aix-builtin-mapping.c#L16

Reply:
> For compile-time evaluation, the APFloats associated with long double on AIX have the IEEE double format (and the evaluation works fine).

Does the evaluation produce the same results as would happen at runtime despite the different format?

Reply:
> Does the evaluation produce the same results as would happen at runtime despite the different format?

There is no different format. The default format for the `long double` type on AIX (and the only one supported by Clang for AIX) is IEEE double.

The IR generated for basic `long double` operations might clarify: https://godbolt.org/z/W4jqr6Kva

Reply:
> With this change (special casing frexpl for AIX) the IR produced is: `call { double, i32 } @llvm.frexp.f64.i32(double 1.234500e+02)`
> 
> without this change the IR is: `store double 0.000000e+00, ptr %returnValue, align 8`
> 
> This doesn't seem correct to me.

Ignoring considerations around storing to `DummyInt`, the original test code was:
```
  returnValue = __builtin_frexpl(0.0L, &DummyInt);
```
I don't know where the `123.45` came from. That the result is zero for an input of zero is correct.

Reply:
> I was referring to https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGBuiltin.cpp#L3543.

That affects the IR generated by Clang when Clang _does not_ constant-evaluate (and was motivated by backend issues).

Reply:
> > With this change (special casing frexpl for AIX) the IR produced is: `call { double, i32 } @llvm.frexp.f64.i32(double 1.234500e+02)`
> > without this change the IR is: `store double 0.000000e+00, ptr %returnValue, align 8`
> > This doesn't seem correct to me.
> 
> Ignoring considerations around storing to `DummyInt`, the original test code was:
> 
> ```
>   returnValue = __builtin_frexpl(0.0L, &DummyInt);
> ```
> 
> I don't know where the `123.45` came from. That the result is zero for an input of zero is correct; however, the constant-folding in the context of that test would undermine its intent (see check that an appropriate target runtime function is called).

Sorry. The output is:   `call { double, i32 } @llvm.frexp.f64.i32(double 0.000000e+00)`. That was a mistake of my part.